examples/error_estimates_forces.jl

```diff
@@ -0,0 +1,184 @@
+# # Practical error bounds for the forces
+#
+# This is a simple example showing how to compute error estimates for the forces on a
+# ``{\rm TiO}_2`` molecule, from which we can either compute asymptotically valid error
+# bounds or increase the precision on the computation of the forces.
+#
+# The strategy we follow is described with more details in[^CDKL2021] and we
+# will use in comments the density matrices framework. We will also needs
+# operators and functions from
+# [`src/scf/newton.jl`](https://dftk.org/blob/master/src/scf/newton.jl).
+#
+# [^CDKL2021]:
+#     E. Cancès, G. Dusson, G. Kemlin, and A. Levitt
+#     *Practical error bounds for properties in plane-wave electronic structure
+#     calculations* Preprint, 2021. [arXiv](https://arxiv.org/abs/2111.01470)
+using DFTK
+using LinearAlgebra
+using ForwardDiff
+using LinearMaps
+using IterativeSolvers
+
+# ## Setup
+# We setup manually the ``{\rm TiO}_2`` configuration from
+# [Materials Project](https://materialsproject.org/materials/mp-2657/).
+Ti = ElementPsp(:Ti, psp=load_psp("hgh/lda/ti-q4.hgh"))
+O = ElementPsp(:O, psp=load_psp("hgh/lda/o-q6.hgh"))
+atoms = [Ti => [[0.5, 0.5, 0.5],
+                [0.0, 0.0, 0.0]],
+         O  => [[0.19542, 0.80458, 0.5],
+                [0.80458, 0.19542, 0.5],
+                [0.30458, 0.30458, 0.0],
+                [0.69542, 0.69542, 0.0]]]
+lattice = [[8.79341  0.0      0.0];
+           [0.0      8.79341  0.0];
+           [0.0      0.0      5.61098]];
+
+# We apply a small displacement to one of the ``\rm Ti`` atoms to get nonzero forces.
+el, pos = atoms[1]
+atoms[1] = Pair(el, pos .+ [[0.22, -0.28, 0.35] / 5, [0, 0, 0]]);
+# We build a model with one k-point only, not too high `Ecut_ref` and small tolerance to
+# limit computational time. These parameters can be increased for more precise
+# results.
+model = model_LDA(lattice, atoms)
+kgrid = [1, 1, 1]
+Ecut_ref = 50
+basis_ref = PlaneWaveBasis(model; Ecut=Ecut_ref, kgrid=kgrid)
+tol = 1e-6;
+
+# ## Computations
+# We compute the reference solution ``P_*`` from which we will compute the
+# references forces.
+scfres_ref = self_consistent_field(basis_ref, tol=tol, callback=info->nothing)
+ψ_ref = DFTK.select_occupied_orbitals(basis_ref, scfres_ref.ψ)
```

Comment:
returning the occupations without the virtual states would suit the best in my opinion (as it is computed just below that)

Reply:
then we have to modify a few files to adapt the output but that's not a big deal

Reply:
can you also allow a tolerance parameter to specify when an orbital is considered "empty" (than this function can also be useful if we have temperature).

Reply:
yup I can do that

Reply:
Just a question about this tolerance stuff for temperature: will it be an issue if the chosen tolerance is such that the returned occupation array has for instance 3 orbitals considered as occupied on one k-point and 4 on another ?

I'm thinking of the example in the doc https://docs.dftk.org/stable/examples/metallic_systems/ where the occupation array is
```
 [1.9999999999077942, 1.9999975862813721, 0.004016889271919304, 3.0004096949872443e-15, 1.1115968307815484e-18, 1.1115046659398737e-18, 7.978744566129872e-19, 7.978122691422918e-19, 3.295359361101644e-22]
 [1.9999997482868157, 1.9949327490941309, 0.0017168724035947042, 1.892360555738378e-6, 8.115209558888162e-8, 9.359437682919286e-13, 6.491292093256281e-14, 1.2325425611911432e-16, 1.0641105741057316e-16]
 [1.3387789742542426, 1.3344369226783412, 1.3344251909990323, 0.0002626279766479971, 6.750936578704564e-5, 6.750709969200984e-5, 8.293218866624051e-13, 1.505956414918229e-13, 1.5053101610349944e-13]
```
and thus taking the tolerance to differentiate between virtual and occupied equal to 1e-4 would select 3 orbitals in the two first k-points and 4 in the third k-point.

Reply:
yeah this doesn't really make much sense in the metallic case (or at least it's pretty tricky). Or are you concerned about unconverged bands @mfherbst ?

Reply:
Yep, we had included this test in `select_occupied_orbitals` 
https://github.com/JuliaMolSim/DFTK.jl/blob/fedc720dab2d194b30d468501acd0f04bd4dd3d6/src/orbitals.jl#L5

Reply:
Not really. I only want a way to distinguish occupied from unoccupied bands also for metals, that's all. So indeed I would also vote for  just returning the max (so 4 bands everywhere).

Reply:
Yes, that's a solution. It won't change anything for insulators, maybe we rise a warning with metals ?

Reply:
but it doesn't really make sense to talk about unoccupied bands for metals. Ideally we should just change self_consistent_field to accept as input a max occupancy and adjust the number of bands dynamically, which would make this irrelevant. We still do have the ambiguity (both in the insulator and metallic case) that it is sometimes useful (and sometimes not) to have partially converged bands. Maybe we should have psi_extra, occupation_extra fields in scfres, so that psi and occupation refer only to converged, occupied (according to the user-given criterion) bands and the extra bands are partially converged bands included to make it easier to converge

Reply:
ok than don't call it unoccupied bands. The point is just to strip off those bands with small occupation, that can be ignored.

I agree that the best solution would be to have a dynamic adjustment of the bands during the SCF (https://github.com/JuliaMolSim/DFTK.jl/issues/562). Actually the "extra" thing you mention sort of already exists, as `n_ep_extra` is always returned, so you can just drop the last few off, both for metals and insulators.

Reply:
From what I understand, the idea is just to have this tool to remove bands we don't want to use (it's not even an exported function), so for the moment maybe we can just use something like that
```
function select_occupied_orbitals(ψ, occupation; threshold=0.0)
    N = maximum(length.([filter(x->x>threshold, occk) for occk in occupation])) 
   [ψk[:, 1:N] for ψk in ψ], [occk[1:N] for occk in occupation]
end
```

Reply:
Better use `findfirst` instead of `filter` and `length`. You can assume occupations are partially ordered.

Reply:
Or if you want to be super fancy, then `findlast`, because very likely there will be fewer orbitals stripped off than kept.

Reply:
@antoine-levitt would this suits you, at least before having dynamic adjustment in the SCF ?

Reply:
I don't get the problem with having different number of bands for different kpoints?

Reply:
I think that could break stuff in DFTK as we assume the number to be identical at places, that's my only concern.

Reply:
Yes, that was my concern too.

Reply:
Do we? We shouldn't: in the metallic case we want to be able to have a different number of bands per kpoint. Psi already has a varying number of rows, so having a varying number of columns is OK. `unpack_psi` looks like it does assume but in fact doesn't: we should remove `n_bands` there (@gkemlin can you do it while you're at it?)

Reply:
I'm all for removing that constraint. That will also help with https://github.com/JuliaMolSim/DFTK.jl/issues/562.

Reply:
I'm not sure we even have that constraint currently, but yes a more systematic review would be nice

Reply:
Ok, so I changed the function to
```
function select_occupied_orbitals(basis, ψ, occupation; threshold=0.0)
    N = [findlast(x -> x > threshold, occk) for occk in occupation]
    selected_ψ = [ψk[:, 1:N[ik]] for (ik,ψk) in enumerate(ψ)]
    selected_occ = [occk[1:N[ik]] for (ik,occk) in enumerate(occupation)]

    # if we have an insulator, sanity check that the orbitals we kept are the
    # occupied ones
    if threshold == 0.0
        model = basis.model
        filled_occ = filled_occupation(model)
        n_spin = model.n_spin_components
        n_bands = div(model.n_electrons, n_spin * filled_occ, RoundUp)
        @assert n_bands == size(selected_ψ[1], 2)
    end

    selected_ψ, selected_occ
end
```
and changed the output in the place where it was used before.